x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/ranges/range_editor.tsx

```diff
@@ -0,0 +1,502 @@
+                    <EuiText size="s">{gteAppendLabel}</EuiText>
+                  </EuiToolTip>
+                }
+                fullWidth
+                compressed
+                placeholder={FROM_PLACEHOLDER}
+                isInvalid={isValidNumber(from) && !isRangeWithin(tempRange)}
+              />
+            </EuiFlexItem>
+            <EuiFlexItem grow={false}>
+              <EuiIcon type="sortRight" color="subdued" />
+            </EuiFlexItem>
+            <EuiFlexItem>
+              <EuiFieldNumber
+                value={isFinite(to) ? Number(to) : ''}
+                onChange={({ target }) => {
+                  const newRange = {
+                    ...tempRange,
+                    to: target.value ? Number(target.value) : -Infinity,
+                  };
+                  setTempRange(newRange);
+                  saveRangeIfValid(newRange);
+                }}
+                prepend={
+                  <EuiToolTip content={ltTooltipContent}>
+                    <EuiText size="s">{ltPrependLabel}</EuiText>
+                  </EuiToolTip>
+                }
+                fullWidth={true}
+                compressed={true}
+                placeholder={TO_PLACEHOLDER}
+                isInvalid={isValidNumber(to) && !isRangeWithin(tempRange)}
+              />
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        </EuiFormRow>
+        <EuiFormRow fullWidth>
+          <EuiFieldText
+            compressed
+            value={safeLabel}
+            prepend={i18n.translate('xpack.lens.indexPattern.ranges.customIntervalPopoverLabel', {
+              defaultMessage: 'Label',
+            })}
+            onChange={({ target }) => {
+              const newRange = {
+                ...tempRange,
+                label: target.value,
+              };
+              setTempRange(newRange);
+              saveRangeIfValid(newRange);
+            }}
+          />
+        </EuiFormRow>
+      </EuiForm>
+    </EuiPopover>
+  );
+};
+
+const AdvancedRangeEditor = ({
+  ranges,
+  setRanges,
+  onToggleEditor,
+}: {
+  ranges: RangeType[];
+  setRanges: (newRanges: RangeType[]) => void;
+  onToggleEditor: () => void;
+}) => {
+  // use a local state to store ids with range objects
+  const [localRanges, setLocalRanges] = useState<LocalRangeType[]>(() =>
+    ranges.map((range) => ({ ...range, id: generateId() }))
+  );
+  // we need to force the open state of the popover from the outside in some scenarios
+  // so we need an extra state here
+  const [isOpenByCreation, setIsOpenByCreation] = useState(false);
+  const lastIndex = localRanges.length - 1;
+
+  const updateRanges = useCallback(
+    (updatedRanges: LocalRangeType[]) => {
+      setRanges(updatedRanges.map(({ id, ...rest }) => ({ ...rest })));
+      setLocalRanges(updatedRanges);
+    },
+    [setRanges]
+  );
+
+  const updateRangesDebounced = React.useMemo(() => debounce(updateRanges, 256), [updateRanges]);
```

Comment:
I recommend the `useDebounce` hook from `react-use` which we have available internally. Handles the edge cases a little better, like cancelling the timeouts.